manager/scheduler/indexed_node_heap.go

```diff
@@ -96,6 +99,47 @@ func (nh *nodeHeap) findMin(meetsConstraints func(*NodeInfo) bool, scanAllNodes
 	return nh.searchHeapToFindMin(meetsConstraints)
 }
 
+type sorter struct {
+	nodes    []*NodeInfo
+	lessFunc func(*NodeInfo, *NodeInfo) bool
+}
+
+func (s sorter) Len() int {
+	return len(s.nodes)
+}
+
+func (s sorter) Swap(i, j int) {
+	s.nodes[i], s.nodes[j] = s.nodes[j], s.nodes[i]
+}
+
+func (s sorter) Less(i, j int) bool {
+	return s.lessFunc(s.nodes[i], s.nodes[j])
+}
+
+// findBestNodes returns n nodes (or < n if fewer nodes are available) that
+// rank best (lowest) according to the sorting function.
+func (nh *nodeHeap) findBestNodes(n int, meetsConstraints func(*NodeInfo) bool, nodeLess func(*NodeInfo, *NodeInfo) bool) []*NodeInfo {
+	var nodes []*NodeInfo
+	for i := 0; i < len(nh.heap); i++ {
+		if meetsConstraints(&nh.heap[i]) {
+			nodes = append(nodes, &nh.heap[i])
+		}
+	}
+
+	// TODO(aaronl): Use quickselect instead of sorting. Also, once we
+	// switch to quickselect, avoid checking constraints on every node, and
+	// instead do it lazily as part of the quickselect process (treat any
+	// node that doesn't meet the constraints as greater than all nodes that
+	// do, and cache the result of the constraint evaluation).
+	sort.Sort(sorter{nodes: nodes, lessFunc: nodeLess})
+
+	if len(nodes) < n {
```

Comment:
I think it's actually necessary to sort the list whether we're returning all the nodes or not. The caller will start with the "best" node at the start of the slice and add tasks to it until that node is no longer "best". The current code isn't doing this yet. I'll fix that.